# Import Libraries and Load Dataset

In [1]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install ipypivot

Note: you may need to restart the kernel to use updated packages.


In [3]:
# process
import numpy as np
import pandas as pd

# viz
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# text process
import re

# viz maps
import folium
from folium import plugins

#pivottable
from pivottablejs import pivot_ui

In [4]:
df = pd.read_csv("Data Faskes BPJS 2019.csv", header=0)
df.head()

NoLink                  Provinsi  \
0       0  Nanggroe Aceh Darussalam   
1       0  Nanggroe Aceh Darussalam   
2       0  Nanggroe Aceh Darussalam   
3       0  Nanggroe Aceh Darussalam   
4       0  Nanggroe Aceh Darussalam   

                                             KotaKab  \
0  Kode Faskes dan Alamat Rumah Sakit BPJS di Kot...   
1  Kode Faskes dan Alamat Rumah Sakit BPJS di Kot...   
2  Kode Faskes dan Alamat Rumah Sakit BPJS di Kot...   
3  Kode Faskes dan Alamat Rumah Sakit BPJS di Kot...   
4  Kode Faskes dan Alamat Rumah Sakit BPJS di Kot...   

                                                Link   TipeFaskes No  \
0  https://lovia.life/id/health/bpjs/area/kota-la...  Rumah Sakit  1   
1  https://lovia.life/id/health/bpjs/area/kota-la...  Rumah Sakit  2   
2  https://lovia.life/id/health/bpjs/area/kota-la...  Rumah Sakit  3   
3  https://lovia.life/id/health/bpjs/area/kota-la...  Rumah Sakit  4   
4  https://lovia.life/id/health/bpjs/area/kota-la...    Puskesmas  1   

  KodeFaskes                                         NamaFaskes  \
0   0001R001  RSU Cut Nyak Dhien\r\n                    \r\n...   
1   0001R004  RSU Cut Meutia Langsa\r\n                    \...   
2   0105R001  RSUD Langsa\r\n                    \r\n       ...   
3   0105R013  RSU Ummi \r\n                    \r\n         ...   
4   00010001  Langsa Lama\r\n                    \r\n       ...   

                                    LatLongFaskes  \
0  http://maps.google.co.id/?q=4.488058,97.947963   
1  http://maps.google.co.id/?q=4.488088,97.947781   
2  http://maps.google.co.id/?q=4.472208,97.975533   
3  http://maps.google.co.id/?q=4.470376,97.991915   
4  http://maps.google.co.id/?q=4.478172,97.949988   

                   AlamatFaskes               TelpFaskes  
0    Jl. Tm Bahrum No. 1 Langsa    \r\n 0641-0621039\r\n  
1   Jl.Garuda Kebun Baru Langsa   \r\n 0641- 4840076\r\n  
2       Jln.A.Yani No. 1 Langsa      \r\n 0641-22051\r\n  
3  Jln. Prof. A. Madjid Ibrahim      \r\n 0641-22886\r\n  
4       Gampong Meurandeh Dayah  \r\n 0641-064121218\r\n

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29157 entries, 0 to 29156
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   NoLink         29157 non-null  int64 
 1   Provinsi       29157 non-null  object
 2   KotaKab        29157 non-null  object
 3   Link           29157 non-null  object
 4   TipeFaskes     29157 non-null  object
 5   No             29157 non-null  object
 6   KodeFaskes     29157 non-null  object
 7   NamaFaskes     29157 non-null  object
 8   LatLongFaskes  29157 non-null  object
 9   AlamatFaskes   29157 non-null  object
 10  TelpFaskes     29157 non-null  object
dtypes: int64(1), object(10)
memory usage: 2.4+ MB


In [6]:
df.columns

Index(['NoLink', 'Provinsi', 'KotaKab', 'Link', 'TipeFaskes', 'No',
       'KodeFaskes', 'NamaFaskes', 'LatLongFaskes', 'AlamatFaskes',
       'TelpFaskes'],
      dtype='object')

# Preprocessing Data

### Set Kota/Kab

In [7]:
def setspace(space):
    word = " ".join(space.split())
    return word

def setkotakab(kotakab):
    word1 = " ".join(kotakab.split())
    word2 = word1.replace('Kode Faskes dan Alamat Rumah Sakit BPJS di','')
    return word2

In [8]:
df['KotaKab'] = df.KotaKab.apply(setkotakab)
df['NamaFaskes'] = df.NamaFaskes.apply(setspace)
df['TelpFaskes'] = df.TelpFaskes.apply(setspace)
df.drop(['Link', 'NoLink'], axis=1, inplace=True)

df.head(10)

Provinsi       KotaKab                 TipeFaskes No  \
0  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  1   
1  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  2   
2  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  3   
3  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  4   
4  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  1   
5  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  2   
6  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  3   
7  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  4   
8  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  5   
9  Nanggroe Aceh Darussalam   Kota Langsa  Dokter Praktik Perorangan  -   

  KodeFaskes             NamaFaskes  \
0   0001R001     RSU Cut Nyak Dhien   
1   0001R004  RSU Cut Meutia Langsa   
2   0105R001            RSUD Langsa   
3   0105R013               RSU Ummi   
4   00010001            Langsa Lama   
5   00010801            Langsa Baro   
6   00010802           Langsa Barat   
7   00010803            Langsa Kota   
8   00011801           Langsa Timur   
9          -                      -   

                                    LatLongFaskes  \
0  http://maps.google.co.id/?q=4.488058,97.947963   
1  http://maps.google.co.id/?q=4.488088,97.947781   
2  http://maps.google.co.id/?q=4.472208,97.975533   
3  http://maps.google.co.id/?q=4.470376,97.991915   
4  http://maps.google.co.id/?q=4.478172,97.949988   
5  http://maps.google.co.id/?q=4.474268,97.957391   
6  http://maps.google.co.id/?q=4.471487,97.989911   
7  http://maps.google.co.id/?q=4.471071,97.974053   
8  http://maps.google.co.id/?q=4.434317,98.017183   
9                                               -   

                   AlamatFaskes      TelpFaskes  
0    Jl. Tm Bahrum No. 1 Langsa    0641-0621039  
1   Jl.Garuda Kebun Baru Langsa   0641- 4840076  
2       Jln.A.Yani No. 1 Langsa      0641-22051  
3  Jln. Prof. A. Madjid Ibrahim      0641-22886  
4       Gampong Meurandeh Dayah  0641-064121218  
5         Jl. Lilawangsa Langsa   0000-00000000  
6     Jl.Prof.A. Madjid Ibrahim  0000-064120282  
7        Jl. H.Agussalim No.10   0641-064121912  
8         Medan-Banda Aceh Km 7   0821-65824360  
9                             -               -

### Clear latitude and longitude

In [9]:
def searchLatLong(row):
    str = re.search('(-?([0-9]{1}|[0-9]0|[1-8]{1,2}).[0-9]{1,6},(-?(1[0-8]{1,2}|9[1-9]{1}).[1-9]{1,6}))', row)
    if str:
        return str.group()
    return np.NaN

df['LatLongFaskes'] = df['LatLongFaskes'].apply(searchLatLong)
df.head(10)

Provinsi       KotaKab                 TipeFaskes No  \
0  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  1   
1  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  2   
2  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  3   
3  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  4   
4  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  1   
5  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  2   
6  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  3   
7  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  4   
8  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  5   
9  Nanggroe Aceh Darussalam   Kota Langsa  Dokter Praktik Perorangan  -   

  KodeFaskes             NamaFaskes       LatLongFaskes  \
0   0001R001     RSU Cut Nyak Dhien  4.488058,97.947963   
1   0001R004  RSU Cut Meutia Langsa  4.488088,97.947781   
2   0105R001            RSUD Langsa  4.472208,97.975533   
3   0105R013               RSU Ummi  4.470376,97.991915   
4   00010001            Langsa Lama  4.478172,97.949988   
5   00010801            Langsa Baro  4.474268,97.957391   
6   00010802           Langsa Barat  4.471487,97.989911   
7   00010803            Langsa Kota     4.471071,97.974   
8   00011801           Langsa Timur                 NaN   
9          -                      -                 NaN   

                   AlamatFaskes      TelpFaskes  
0    Jl. Tm Bahrum No. 1 Langsa    0641-0621039  
1   Jl.Garuda Kebun Baru Langsa   0641- 4840076  
2       Jln.A.Yani No. 1 Langsa      0641-22051  
3  Jln. Prof. A. Madjid Ibrahim      0641-22886  
4       Gampong Meurandeh Dayah  0641-064121218  
5         Jl. Lilawangsa Langsa   0000-00000000  
6     Jl.Prof.A. Madjid Ibrahim  0000-064120282  
7        Jl. H.Agussalim No.10   0641-064121912  
8         Medan-Banda Aceh Km 7   0821-65824360  
9                             -               -

In [10]:
# Split LatLongFaskes menjadi dua kolom dan ubah ke numerik
df[['Latitude', 'Longitude']] = df['LatLongFaskes'].str.split(',', expand=True).apply(pd.to_numeric, errors='coerce')

# Filter nilai yang valid
df['Latitude'] = np.where((-90.0 <= df['Latitude']) & (df['Latitude'] <= 90.0), df['Latitude'], np.NaN)
df['Longitude'] = np.where((-180.0 <= df['Longitude']) & (df['Longitude'] <= 180.0), df['Longitude'], np.NaN)

# Hapus kolom LatLongFaskes
df.drop(['LatLongFaskes'], axis=1, inplace=True)

df.head(10)

Provinsi       KotaKab                 TipeFaskes No  \
0  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  1   
1  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  2   
2  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  3   
3  Nanggroe Aceh Darussalam   Kota Langsa                Rumah Sakit  4   
4  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  1   
5  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  2   
6  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  3   
7  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  4   
8  Nanggroe Aceh Darussalam   Kota Langsa                  Puskesmas  5   
9  Nanggroe Aceh Darussalam   Kota Langsa  Dokter Praktik Perorangan  -   

  KodeFaskes             NamaFaskes                  AlamatFaskes  \
0   0001R001     RSU Cut Nyak Dhien    Jl. Tm Bahrum No. 1 Langsa   
1   0001R004  RSU Cut Meutia Langsa   Jl.Garuda Kebun Baru Langsa   
2   0105R001            RSUD Langsa       Jln.A.Yani No. 1 Langsa   
3   0105R013               RSU Ummi  Jln. Prof. A. Madjid Ibrahim   
4   00010001            Langsa Lama       Gampong Meurandeh Dayah   
5   00010801            Langsa Baro         Jl. Lilawangsa Langsa   
6   00010802           Langsa Barat     Jl.Prof.A. Madjid Ibrahim   
7   00010803            Langsa Kota        Jl. H.Agussalim No.10    
8   00011801           Langsa Timur         Medan-Banda Aceh Km 7   
9          -                      -                             -   

       TelpFaskes  Latitude  Longitude  
0    0641-0621039  4.488058  97.947963  
1   0641- 4840076  4.488088  97.947781  
2      0641-22051  4.472208  97.975533  
3      0641-22886  4.470376  97.991915  
4  0641-064121218  4.478172  97.949988  
5   0000-00000000  4.474268  97.957391  
6  0000-064120282  4.471487  97.989911  
7  0641-064121912  4.471071  97.974000  
8   0821-65824360       NaN        NaN  
9               -       NaN        NaN

# Analizing

### Provinces that have Health Facilities with BPJS

In [11]:
province = df.Provinsi.unique().tolist()
print(province)
print('\nTotal province: ', len(province))

['Nanggroe Aceh Darussalam', 'Sumatera Utara', 'Sumatera Barat', 'Riau', 'Jambi', 'Sumatera Selatan', 'Bengkulu', 'Lampung', 'Kep. Bangka Belitung', 'Dki Jakarta', 'Jawa Barat', 'Jawa Tengah', 'D I Yogyakarta', 'Jawa Timur', 'Banten', 'Bali', 'Nusa Tenggara Barat', 'Nusa Tenggara Timur', 'Kalimantan Barat', 'Kalimantan Tengah', 'Kalimantan Selatan', 'Kalimantan Timur', 'Sulawesi Utara', 'Sulawesi Tengah', 'Sulawesi Selatan', 'Sulawesi Tenggara', 'Gorontalo', 'Maluku', 'Maluku Utara', 'Papua', 'Papua Barat', 'Sulawesi Barat', 'Kepulauan Riau', 'Kalimantan Utara']

Total province:  34


### City/Regency with BPJS Health Facilities

In [12]:
city_regency = df['KotaKab'].unique().tolist()
print(city_regency)
print('\nTotal: ', len(city_regency))

[' Kota Langsa', ' Kab. Gayo Lues', ' Kab. Aceh Selatan', ' Kab. Aceh Tenggara', ' Kab. Aceh Timur', ' Kab. Aceh Tengah', ' Kab.Aceh Barat', ' Kab. Aceh Besar', ' Kab. Pidie', ' Kab. Aceh Tamiang', ' Kab. Aceh Utara', ' Kab. Bireun', ' Kota Lhokseumawe', ' Kab. Aceh Singkil', ' Kab. Simeulue', ' Kota Banda Aceh', ' Kota Sabang', ' Kab. Aceh Jaya', ' Kab. Nagan Raya', ' Kab. Aceh Barat Daya', ' Kab. Bener', ' Kab. Pidie Jaya', ' Kab. Subulussalam', ' Kab. Nias', ' Kab. Mandailing Natal', ' Kab. Tapanuli Selatan', ' Kab. Tapanuli Tengah', ' Kab. Tapanuli Utara', ' Kab. Toba Samosir', ' Kab. Labuhanbatu', ' Kab. Asahan', ' Kab. Simalungun', ' Kab. Dairi', ' Kab. Karo', ' Kab. Deli Serdang', ' Kab. Langkat', ' Kota Sibolga', ' Kota Tanjung Balai', ' Kota Pematang Siantar', ' Kota Tebing Tinggi', ' Kota Medan', ' Kota Binjai', ' Kab. Serdang Bedagai', ' Kab. Samosir', ' Kab. Pakpak Bharat', ' Kab. Nias Selatan', ' Kab. Humbang Hasundutan', ' Kota Padang Sidempuan', ' Kab. Batu Bara', ' Kab.

### Data to be displayed in the visualization map

In [13]:
total_data_count = df.shape[0]
displayed_data = df.dropna().shape[0]
data_not_displayed = df[df['Latitude'].isnull() | df['Longitude'].isnull()].shape[0]

print('\nTotal facility data: ', total_data_count)
print('\nTotal facilities that can be displayed on the map:', displayed_data)
print('\nTotal facilities that cannot be displayed on the map:', data_not_displayed)


Total facility data:  29157

Total facilities that can be displayed on the map: 16499

Total facilities that cannot be displayed on the map: 12658


# Data Viz

### Geographical visualization of 16499 BPJS health facilities

#### Visualization of geographic data with folium FastMarkCluster

In [14]:
data_map = df.dropna().info()
data_map

<class 'pandas.core.frame.DataFrame'>
Index: 16499 entries, 0 to 29149
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Provinsi      16499 non-null  object 
 1   KotaKab       16499 non-null  object 
 2   TipeFaskes    16499 non-null  object 
 3   No            16499 non-null  object 
 4   KodeFaskes    16499 non-null  object 
 5   NamaFaskes    16499 non-null  object 
 6   AlamatFaskes  16499 non-null  object 
 7   TelpFaskes    16499 non-null  object 
 8   Latitude      16499 non-null  float64
 9   Longitude     16499 non-null  float64
dtypes: float64(2), object(8)
memory usage: 1.4+ MB


In [15]:
maps_dataset = df.dropna()
rome_lat, rome_lng = -6.200000, 106.816666       # jakarta

# init the folium map object
maps = folium.Map(location=[rome_lat, rome_lng], zoom_start=5)

# input all points from the dataset file into a map object using FastMarkerCluster
maps.add_child(plugins.FastMarkerCluster(maps_dataset[['Latitude', 'Longitude']].values.tolist()))

maps

#### Visualization of numbers per city/district with available latitude and longitude

In [16]:
# 5 cities or districts with the largest number of BPJS facilities
city_regency = df['KotaKab'].value_counts().rename_axis('city_regrency').reset_index(name='Total')
city_regency.head()

city_regrency  Total
0    Kota Surabaya    364
1    Kota Semarang    305
2       Kab. Bogor    298
3    Kota Makassar    292
4   Kota Palembang    288

In [17]:
# Retrieve latitude and longitude data per city/district
kabkota = []
prov = []
lati = []
longi = []

df1 = df.dropna()

for prv, kk, ltk, lnk in zip(df1['Provinsi'], 
                             df1['KotaKab'], 
                             df1['Latitude'], 
                             df1['Longitude']):
    if (not(kk in kabkota) and not(ltk == np.NaN) and not(lnk == np.NaN)):
        prov.append(prv)
        kabkota.append(kk)
        lati.append(ltk)
        longi.append(lnk)

print(len(kabkota))

data_fas = pd.DataFrame(list(zip(prov, kabkota, lati, longi)),
                        columns=['Provinsi', 'city_regrency', 'Latitude', 'Langitude'])
data_fas.head()             

460


Provinsi        city_regrency  Latitude  Langitude
0  Nanggroe Aceh Darussalam          Kota Langsa  4.488058  97.947963
1  Nanggroe Aceh Darussalam       Kab. Gayo Lues  4.059445  97.463333
2  Nanggroe Aceh Darussalam    Kab. Aceh Selatan  3.111916  97.368642
3  Nanggroe Aceh Darussalam   Kab. Aceh Tenggara  3.496266  97.800000
4  Nanggroe Aceh Darussalam      Kab. Aceh Timur  4.961196  97.762670

In [18]:
merged = pd.merge(data_fas, city_regency, on='city_regrency')
merged.count()

Provinsi         460
city_regrency    460
Latitude         460
Langitude        460
Total            460
dtype: int64

In [19]:
# Create map
bpjs_map = folium.Map(location=[10, -20], zoom_start=2.3, tiles='Stamen Toner', attr='Map data &copy; Stamen Toner')

for lat, lon, prov, kotakab, jml in zip(merged['Latitude'], merged['Langitude'], merged['Provinsi'], merged['city_regrency'], merged['Total']):
    folium.CircleMarker([lat, lon],
                        radius=10,
                        popup = ('<strong>Provinsi</strong>: ' + str(prov) + '<br>'
                                '<strong>Kota/Kab</strong>: ' + str(kotakab) + '<br>'
                                '<strong>Jumlah</strong>: ' + str(jml)),
                        color='blue',  # Warna garis diubah menjadi pink
                        fill_color='yellow',  # Warna isian diubah menjadi pink
                        fill_opacity=0.3 ).add_to(bpjs_map)

bpjs_map


### Data by province

In [20]:
data_sorted = pd.DataFrame({'Provinsi':merged['city_regrency'], 'Total':merged['Total']})
data_sorted = data_sorted.sort_values(by=['Total'], ascending=False)

data_sorted.head()

Provinsi  Total
243    Kota Surabaya    364
201    Kota Semarang    305
151       Kab. Bogor    298
103   Kota Palembang    288
169     Kota Bandung    276

## Visualization of all data

### City/Regency with the highest number of health facilities with BPJS

In [21]:
df.Provinsi.value_counts()

Provinsi
Jawa Tengah                 3688
Jawa Barat                  3683
Jawa Timur                  3548
Sumatera Utara              1628
Sulawesi Selatan            1155
Dki Jakarta                  995
Sumatera Selatan             993
Banten                       913
Lampung                      784
Nanggroe Aceh Darussalam     776
Bali                         765
Riau                         729
Nusa Tenggara Timur          722
Sumatera Barat               721
Papua                        712
Kalimantan Selatan           671
Kalimantan Timur             634
Sulawesi Utara               543
D I Yogyakarta               534
Kalimantan Barat             531
Sulawesi Tenggara            500
Jambi                        446
Kalimantan Tengah            418
Nusa Tenggara Barat          393
Bengkulu                     390
Sulawesi Tengah              386
Maluku                       334
Kepulauan Riau               319
Papua Barat                  278
Maluku Utara                 253
G

In [22]:
df.KotaKab.value_counts()

KotaKab
 Kota Surabaya                364
 Kota Semarang                305
 Kab. Bogor                   298
 Kota Makassar                292
 Kota Palembang               288
                             ... 
 Kab.Bolaang Mongondow Tmr     12
 Kab. Banggai Laut             12
 Kab. Mamberamo Tengah         11
 Kab. Tana Tidung              11
 Kab. Manokwari Selatan        10
Name: count, Length: 512, dtype: int64

In [23]:
df.TipeFaskes.value_counts()

TipeFaskes
Puskesmas                    9913
Klinik Pratama               6597
Klinik Utama                 5375
Apotek                       2982
Rumah Sakit                  2249
Dokter Gigi                  1410
Dokter Praktik Perorangan     631
Name: count, dtype: int64

## Pivotable in Python

### Total Faskes in Prov

In [24]:
from pivottablejs import pivot_ui

In [25]:
pivot_ui(df)

### Faskes in Prov and kab/Kota

In [26]:
pivot_ui(df)

### Type Faskes in Prov and kab/Kota

In [27]:
pivot_ui(df)

In [31]:
df.to_excel('FasKes BPJS 2019.xlsx', index=False)